torch/csrc/jit/script/compiler.cpp

```diff
@@ -494,14 +491,48 @@ struct to_ir {
     auto* false_block = n->addBlock();
 
     // Emit both blocks once to get the union of all mutated values
-    std::unordered_set<std::string> mutated_parent_values;
-    auto save_true = emitSingleIfBranch(
-        true_block, stmt.trueBranch(), &mutated_parent_values);
-    auto save_false = emitSingleIfBranch(
-        false_block, stmt.falseBranch(), &mutated_parent_values);
+    auto save_true = emitSingleIfBranch(true_block, stmt.trueBranch());
+    auto save_false = emitSingleIfBranch(false_block, stmt.falseBranch());
+
+    // In python, every variable assigned in an if statement escapes
+    // the scope of the if statement (all variables are scoped to the function).
+    // Script is a subset of python: we consider variables to be in scope
+    // as long as there is a definition of the variable along all paths
+    // through the if statemnent
+    // ----
+    // if ...:
+    //   a =
+    // else:
+    //   ...
+    // ... = a  # error, a is not defined along all paths
+    // ----
+    // if ...:
+    //   a =
+    // else:
+    //   a =
+    // ... = a # OK, a is defined along all paths
+    // ----
+    // a = ...
+    // if ...:
+    //   a =
+    // ... = a # OK, a is defined along all paths
+
+
+    std::set<std::string> mutated_variables;
```

Comment:
Any reason to use a `set` and not `unordered_set`?

Actually if we do this, then we don't need to `std::sort` the `vector`, and could equivalently loop over this set, so it might make sense.

Reply:
Well, the use of set was a typo on my part, but your point about sorting is valid, so I guess we'll keep it :)